### PR TITLE
Move Python starter template to TypeScript AppHost and CLI template factory

### PR DESCRIPTION
**Pull request by davidfowl**
## Description

Migrates the `aspire-py-starter` template from the old `dotnet new` template system (DotNetTemplateFactory with C# AppHost) to the new embedded CLI template system (CliTemplateFactory with TypeScript AppHost).

**Motivation:** Narrowing down the toolchains required — by moving to a TypeScript AppHost, the Python starter no longer requires the .NET SDK/C# toolchain for the AppHost project, aligning it with the same pattern as `aspire-ts-starter`.

### What changed

- **New embedded template** at `src/Aspire.Cli/Templating/Templates/py-starter/` with a TypeScript AppHost using `addUvicornApp`, conditional Redis support via `{{#redis}}`/`{{/redis}}` block markers, and a Vite React frontend
- **New `CliTemplateFactory.PythonStarterTemplate.cs`** — factory logic for the Python starter, including Redis conditional processing and `aspire.config.json` package injection
- **New `ConditionalBlockProcessor` utility class** — extracted reusable conditional block processing with 29 unit tests covering comment styles, edge cases, multiple conditions, and realistic template scenarios
- **Removed old template** from `Aspire.ProjectTemplates` (`dotnet new aspire-py-starter` is replaced by `aspire new aspire-py-starter`)
- **Updated registrations** — `aspire-py-starter` moved from `DotNetTemplateFactory` to `CliTemplateFactory` with `--use-redis-cache` and `--localhost-tld` options
- **Updated tests** — `DotNetTemplateFactoryTests` updated to expect py-starter not in dotnet templates; hostname tests removed for py-starter

### Validation

- Build succeeds (both `Aspire.Cli` and `Aspire.ProjectTemplates`)
- 29 `ConditionalBlockProcessorTests` pass
- 12 `DotNetTemplateFactoryTests` pass
- 27 `NewCommandTests` pass (confirms `aspire-py-starter` listed in CLI help)
- E2E tested: `aspire new aspire-py-starter` generates correct output for both `--use-redis-cache false` and `--use-redis-cache true` variants

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Link to `aspire.dev` issue:
      - [New issue](https://github.com/microsoft/aspire.dev/issues/new)
  - [ ] No

### CODE REVIEW

File: src/Aspire.Cli/Templating/CliTemplateFactory.PythonStarterTemplate.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+                    var ports = GenerateRandomPorts();
+                    var hostName = useLocalhostTld ? $"{projectNameLower}.dev.localhost" : "localhost";
+                    var conditions = new Dictionary<string, bool>
+                    {
+                        ["redis"] = useRedisCache,
+                        ["no-redis"] = !useRedisCache,
+                    };
+                    string ApplyAllTokens(string content) => ConditionalBlockProcessor.Process(
+                        ApplyTokens(content, projectName, projectNameLower, aspireVersion, ports, hostName),
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Cli/Templating/CliTemplateFactory.PythonStarterTemplate.cs:66]: This is a pre-existing pattern shared by all CLI templates — both `CliTemplateFactory.TypeScriptStarterTemplate.cs` (line 55) and `CliTemplateFactory.EmptyTemplate.cs` (line 167) use the same `projectName.ToLowerInvariant()` without DNS normalization. This PR just follows the established pattern. A DNS-safe normalization improvement could be done as a follow-up across all templates via the shared `ApplyTokens` method.

File: src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        var startPattern = $"{{{{#{blockName}}}}}";
+        var endPattern = $"{{{{/{blockName}}}}}";
+
+        while (true)
+        {
+            var startIdx = content.IndexOf(startPattern, StringComparison.Ordinal);
+            if (startIdx < 0)
+            {
+                break;
```
Comment (JamesNK) [inline src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs:68]: If a start marker exists but the matching end marker is misspelled or missing, the processor silently returns the content unchanged — leaving raw `{{#name}}` markers in the generated template output. This would be very hard to catch since there's no warning. Consider logging a warning (or throwing) when `startIdx >= 0` but `endIdx < 0`, so template authoring typos don't produce silently broken output.

File: src/Aspire.Cli/Templating/CliTemplateFactory.PythonStarterTemplate.cs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+            useRedisCache = await _interactionService.PromptForSelectionAsync(
+                TemplatingStrings.UseRedisCache_Prompt,
+                [TemplatingStrings.Yes, TemplatingStrings.No],
+                choice => choice,
+                cancellationToken) switch
+            {
+                var choice when string.Equals(choice, TemplatingStrings.Yes, StringComparisons.CliInputOrOutput) => true,
+                var choice when string.Equals(choice, TemplatingStrings.No, StringComparisons.CliInputOrOutput) => false,
+                _ => throw new InvalidOperationException(TemplatingStrings.UseRedisCache_UnexpectedChoice)
```
Comment (JamesNK) [inline src/Aspire.Cli/Templating/CliTemplateFactory.PythonStarterTemplate.cs:142]: This method is a near-duplicate of `ResolveUseLocalhostTldAsync` — same check-tokens-then-prompt-interactively pattern. Both check `parseResult.Tokens` for a specific option name, call `parseResult.GetValue`, fall back to `PromptForSelectionAsync` with Yes/No, and display a confirmation message. Consider extracting a shared helper like `ResolveBooleanOptionAsync(parseResult, option, promptText, confirmationText, cancellationToken)` to reduce duplication, especially since more templates with boolean options will likely follow.

File: src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    /// <param name="conditions">A set of block-name to include/exclude mappings.</param>
+    /// <returns>The processed content with conditional blocks resolved.</returns>
+    internal static string Process(string content, IReadOnlyDictionary<string, bool> conditions)
+    {
+        foreach (var (blockName, include) in conditions)
+        {
+            content = ProcessBlock(content, blockName, include);
+        }
+
+        return content;
```
Comment (JamesNK) [inline src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs:42]: `Process()` iterates `IReadOnlyDictionary<string, bool>` whose enumeration order is not contractually guaranteed. For the current `redis`/`no-redis` usage with non-overlapping blocks this works fine, but the API shape invites future callers to add conditions with nested or overlapping blocks, where the behavior would become silently order-dependent.

Minor suggestion: either document the non-overlapping requirement with a comment here, or accept an `IReadOnlyList<(string Name, bool Include)>` to make ordering explicit.

Comment (JamesNK) [inline src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs:42]: Add a debug assert that there are no conditional blocks in the content once complete? i.e. have a regex check for `{{#ALPHA_CHARACTERS}}` and `{{/ALPHA_CHARACTERS}}` and throw an error if there is any match

**Comment by eerhardt** [comment]
This should use python apphost right? https://github.com/microsoft/aspire/pull/13947

**Comment by davidfowl** [comment]
> This should use python apphost right? https://github.com/microsoft/aspire/pull/13947

It can, but the app has javascript, python *and* C#. We're not shipping the python apphost yet. I suspect that will be a bigger move than adding another TS apphost.

**Comment by davidfowl** [comment]
## Status: Blocked on bundle mode multi-package NuGet restore

The template migration itself is complete and works locally (SDK mode). The E2E test currently only verifies template creation + SDK file generation.

**Blocker:** Bundle mode (native AOT CLI) doesn't include integration assemblies in the generated TypeScript SDK when `aspire.config.json` has **multiple packages**. The `aspire-managed nuget restore` with multiple `--package` args produces a libs directory that only contains core `Aspire.Hosting` — both `Aspire.Hosting.JavaScript` and `Aspire.Hosting.Python` are missing.

Evidence from CI workspace capture:
- Generated `aspire.ts` has 16K lines but zero integration-specific methods (`addUvicornApp`, `addViteApp`, etc.)
- Single-package restore works fine (TS starter test passes with `aspire start`/`aspire stop`)
- Two-package restore produces SDK with only core functions

Once the bundle mode multi-package restore is fixed, the E2E test should be updated to include `aspire run` (see comment in `PythonReactTemplateTests.cs`).

File: src/Aspire.Cli/Templating/CliTemplateFactory.PythonStarterTemplate.cs (hunk 3 above)
Comment (davidfowl) [reply src/Aspire.Cli/Templating/CliTemplateFactory.PythonStarterTemplate.cs:142]: Good suggestion — this is a pre-existing pattern (`ResolveUseLocalhostTldAsync` exists identically in `EmptyTemplate` and `TypeScriptStarterTemplate` too). I'd prefer to extract the shared helper as a follow-up across all templates rather than in this PR.

File: src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs (hunk 4 above)
Comment (davidfowl) [reply src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs:42]: Addressed — added a `<remarks>` block documenting the non-overlapping requirement on the class.

Comment (davidfowl) [reply src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs:42]: Addressed — added a `Debug.Assert` with a source-generated regex (`{{[#/][a-zA-Z][\w-]*}}`) that fires after all conditions are processed in `Process()`. This catches leftover markers from typos or missing condition entries.

File: src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Cli/Templating/ConditionalBlockProcessor.cs:68]: Addressed — ProcessBlock now throws InvalidOperationException when a start marker exists but the matching end marker is missing. Updated the test to verify this.

**Comment by davidfowl** [comment]
## Status Update

**Root cause found**: `aspire run` from workspace root uses the root-level `aspire.config.json` (which has no `packages` section) instead of the project-level config. Filed as #15623.

**Fix**: Aligned the E2E test with the TypeScript starter pattern — `cd` into the project directory before running `aspire start`/`aspire stop`. Also addressed all ConditionalBlockProcessor review feedback from @JamesNK.

Previous status comment about bundle mode multi-package restore is now explained: it was never a NuGet restore bug — the config resolution was reading the wrong file.

**Comment by davidfowl** [comment]
## Status Update

Root cause found: aspire run from workspace root uses the root-level aspire.config.json (which has no packages section) instead of the project-level config. Filed as #15623.

Fix: Aligned the E2E test with the TypeScript starter pattern — cd into the project directory before running aspire start/aspire stop. Also addressed all ConditionalBlockProcessor review feedback from JamesNK.

Previous status comment about bundle mode multi-package restore is now explained: it was never a NuGet restore bug — the config resolution was reading the wrong file.

**Review by maddymontaquila — DISMISSED** [review]
changes requested

File: src/Aspire.Cli/Templating/CliTemplateFactory.cs [hunk 5]
```diff
@@ -37,6 +37,11 @@ internal sealed partial class CliTemplateFactory : ITemplateFactory
         Description = TemplatingStrings.UseLocalhostTld_Description
     };
 
+    private readonly Option<bool?> _useRedisCacheOption = new("--use-redis-cache")
+    {
+        Description = "Use Redis for output caching"
+    };
+
     private readonly ILanguageDiscovery _languageDiscovery;
```
Comment (JamesNK) [inline src/Aspire.Cli/Templating/CliTemplateFactory.cs:45]: The `_localhostTldOption` uses `TemplatingStrings.UseLocalhostTld_Description` but this is hardcoded. The resx already has `UseRedisCache_Description` — should use `TemplatingStrings.UseRedisCache_Description` here.

Also the hardcoded text says "output caching" but the template uses Redis as a general data cache (weather forecast), which doesn't match.

File: src/Aspire.Cli/Templating/CliTemplateFactory.cs [hunk 6]
```diff
@@ -130,7 +135,20 @@ private IEnumerable<ITemplate> GetTemplateDefinitions()
...
+                projectName => $"./{projectName}",
+                cmd =>
+                {
+                    AddOptionIfMissing(cmd, _localhostTldOption);
+                    AddOptionIfMissing(cmd, _useRedisCacheOption);
+                },
+                ApplyPythonStarterTemplateAsync,
+                runtime: TemplateRuntime.Cli,
+                languageId: KnownLanguageId.TypeScript)
         ];
```
Comment (JamesNK) [inline src/Aspire.Cli/Templating/CliTemplateFactory.cs:153]: This is registered with `languageId: KnownLanguageId.TypeScript`, which means the template won't appear in `aspire new` if TypeScript language discovery fails — even if Python is available. The old `DotNetTemplateFactory` registration used `KnownLanguageId.CSharp`. Is this intentional that a "Python starter" is gated on TypeScript/Node.js tooling detection? A user who has Python but not Node.js won't see this template.

File: src/Aspire.Cli/Templating/CliTemplateFactory.cs (hunk 5 above)
Comment (davidfowl) [reply src/Aspire.Cli/Templating/CliTemplateFactory.cs:45]: Addressed — replaced the hardcoded string with `TemplatingStrings.UseRedisCache_Description` from the resx.

File: src/Aspire.Cli/Templating/CliTemplateFactory.cs (hunk 6 above)
Comment (davidfowl) [reply src/Aspire.Cli/Templating/CliTemplateFactory.cs:153]: Yes, this is intentional — the Python starter uses a TypeScript AppHost (that's the whole point of this migration), so it requires Node.js/TypeScript tooling. A user needs both Python (for the app) and Node.js (for the AppHost) installed. This matches how the TS starter works and is the expected developer experience for this template.

Comment (JamesNK) [reply src/Aspire.Cli/Templating/CliTemplateFactory.cs:153]: Maybe add a comment to be clear to people/agents why it's configured this way.

**Review by JamesNK — APPROVED** [review]
